README.md
# NAME

Dancer2::Template::TextTemplate - Text::Template engine for Dancer2

# VERSION

version 1.003

# SYNOPSIS

To use this engine, you may configure [Dancer2](https://metacpan.org/pod/Dancer2) via `config.yml`:

    template: text_template

# DESCRIPTION

This template engine allows you to use [Text::Template](https://metacpan.org/pod/Text::Template) in [Dancer2](https://metacpan.org/pod/Dancer2).

## Configuration

Here are all available options, as you would set them in a `config.yml`, with
their __default__ values:

    template: text_template
    engines:
        text_template:
            caching: 1
            expires: 3600               # in seconds; use 0 to disable
            cache_stringrefs: 1
            delimiters: [ "{", "}" ]
            prepend: |
                use strict;
                use warnings;
            safe: 1
            safe_opcodes: [ ":default", ":load" ]
            safe_disposable: 0

The following sections explain what these options do.

## Global caching - `caching`, `expires`

Contrary to other template engines (like [Template::Toolkit](https://metacpan.org/pod/Template::Toolkit)), where _one_
instance may work on _multiple_ templates, _one_ [Text::Template](https://metacpan.org/pod/Text::Template) instance
is created _for each_ template. Therefore, if:

- you don't use a huge amount of different templates;
- you don't use each template just once;

then it may be interesting to __cache__ Text::Template instances for later use.
Since these conditions seem to be common, this engine uses a cache (_via_
[CHI](https://metacpan.org/pod/CHI)) __by default__.

If you're OK with caching, you should specify a __timeout__ (`expires`) after
which cached Text::Template instances are to be refreshed, since you might
have changed your template sources without restarting Dancer2. By default,
this engine uses `expires: 3600` (one hour). Use `0` to tell it that
templates never expire.

If you don't want any caching, just set `caching` to `0`.

## "String-ref templates" caching - `cache_stringrefs`

Just like with [Dancer2::Template::Toolkit](https://metacpan.org/pod/Dancer2::Template::Toolkit), you can pass templates either as
filenames (for a template file) or string references ("string-refs", which are
dereferenced and used as the template's content). In some cases, you may want
to disable caching for string-refs only: for instance, if you generate a lot
of templates on-the-fly and use them only once, caching them is useless and
fills your cache. You can therefore disable caching _for string-refs only_ by
setting `cache_stringrefs` to `0`.

Note that if you set `caching` to `0`, you don't have _any_ caching, so
`cache_stringrefs` is ignored.

## Custom delimiters - `delimiters`

The `delimiters` option allows you to specify a custom delimiters pair
(opening and closing) for your templates. See the [Text::Template](https://metacpan.org/pod/Text::Template)
documentation for more about delimiters, since this module just pass them to
Text::Template. This option defaults to `{` and `}`, meaning that in `a
{b} c`, `b` (and only `b`) will be interpolated.

## Prepending code - `prepend`

This option specifies Perl code run by Text::Template _before_ evaluating
each template. For instance, with this option's default value, i.e.:

    use strict;
    use warnings FATAL => 'all';

then evaluating the following template:

    you're the { $a + 1 }th visitor!

is the same as evaluating:

    {
        use strict;
        use warnings FATAL => 'all';
        ""
    }you're the { $a + 1 }th visitor!

and thus you'd get:

    Program fragment delivered error
    ``Use of uninitialized value $a in addition (+) [...]

in your template output if you forgot to pass a value for `$a`.

If you don't want anything prepended to your templates, simply give a
non-dying, side-effects-free Perl expression to `prepend`, like `0` or
`""`.

## Running in a [Safe](https://metacpan.org/pod/Safe) - `safe`, `safe_opcodes`, `safe_disposable`

This option (enabled by default) makes your templates to be evaluated in a
[Safe](https://metacpan.org/pod/Safe) compartment, i.e. where some potentially dangerous operations (such as
`system`) are disabled. Note that the same Safe compartment will be used to
evaluate all your templates, unless you explicitly specify `safe_disposable:
1` (one compartment per template _evaluation_).

This Safe uses the `:default` and `:load` opcode sets (see [the Opcode
documentation](https://metacpan.org/pod/Opcode#Predefined-Opcode-Tags)), unless
you specify it otherwise with the `safe_opcodes` option. You can, of course,
mix opcodes and optags, as in:

    safe_opcodes:
        - ":default"
        - "time"

which enables the default opcode set _and_ `time`. But __be careful__: with
the previous example for instance, you don't allow `require`, and thus break
the default value of the `prepend` option (which contains `use`)!

# METHODS

## render( $template, \\%tokens )

Renders the template.

- `$template` is either a (string) filename for the template file or a reference to a string that contains the template.
- `\%tokens` is a hashref for the tokens you wish to pass to [Text::Template](https://metacpan.org/pod/Text::Template) for rendering, as if you were using `Text::Template::fill_in`.

[Carp](https://metacpan.org/pod/Croak)s if an error occurs.

# AUTHOR

Thibaut Le Page <thilp@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Thibaut Le Page.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Dancer2-Template-TextTemplate

{{$NEXT}}

1.003     2014-04-17 19:04:56+02:00 Europe/Paris

    - bump version to make it an official release

1.002     2014-04-15 23:50:09+02:00 Europe/Paris (TRIAL RELEASE)

    - require perl-5.10.0 instead of perl-5.8.9:
      Perl 5.8.9's "warnings" pragma uses "Carp", which calls eval(). Since
      "use warnings" is (rightfully imho) a default in the
      D2::T::TextTemplate's "prepend" option, and since eval() is (even more
      rightfully) forbidden by default in a Safe compartment, I see no other
      solution.

1.001     2014-04-15 05:57:08+02:00 Europe/Paris

    - make changes introduced by 1.0 official
    - remove ExtUtils::MakeMaker dependency

1.0       2014-04-14 02:57:57+02:00 Europe/Paris (TRIAL RELEASE)

    - replace the ":load" optag by the "require", "caller" and "dofile"
      opcodes (perl-5.8.9 version of the Opcode module doesn't define ":load")
    - requires IPC::System::Simple for "use autodie ':all'" in tests
    - add Module::Build
    - fix POD errors

0.2       2014-04-10 00:27:31+02:00 Europe/Paris

    - safe, safe_opcodes and safe_disposable options implemented
    - cache_stringrefs option implemented
    - prepend option implemented
    - fixed a bug in render()'s return value that warns only in perl-5.19.x
    - improved build system (more tests, cpanfile)

0.1       2014-04-09 05:10:53+02:00 Europe/Paris

    - first version, not tested enough


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.015.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Text::Template engine for Dancer2",
  "dist_author" => [
    "Thibaut Le Page <thilp\@cpan.org>"
  ],
  "dist_name" => "Dancer2-Template-TextTemplate",
  "dist_version" => "1.002",
  "license" => "perl",
  "module_name" => "Dancer2::Template::TextTemplate",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "CHI" => 0,
    "Carp" => 0,
    "Dancer2::Core::Role::Template" => 0,
    "Dancer2::Core::Types" => 0,
    "Moo" => 0,
    "MooX::Types::MooseLike::Base" => 0,
    "Safe" => "2.26",
    "Scalar::Util" => 0,
    "Text::Template" => "1.46",
    "namespace::clean" => 0,
    "perl" => "5.010_000",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Temp" => 0,
    "IPC::System::Simple" => 0,
    "Test::API" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "autodie" => 0
  }
);


my %fallback_build_requires = (
  "File::Temp" => 0,
  "IPC::System::Simple" => 0,
  "Module::Build" => "0.3601",
  "Test::API" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.88",
  "autodie" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer2/Template/TextTemplate.pm
package Dancer2::Template::TextTemplate;
# ABSTRACT: Text::Template engine for Dancer2

use 5.010_000;
use strict;
use warnings;

# VERSION

use Carp 'croak';
use Moo;
use Dancer2::Core::Types 'InstanceOf';
use Dancer2::Template::TextTemplate::FakeEngine;
use namespace::clean;

with 'Dancer2::Core::Role::Template';

=head1 SYNOPSIS

To use this engine, you may configure L<Dancer2> via C<config.yml>:

    template: text_template

=head1 DESCRIPTION

This template engine allows you to use L<Text::Template> in L<Dancer2>.

=head2 Configuration

Here are all available options, as you would set them in a C<config.yml>, with
their B<default> values:

    template: text_template
    engines:
        text_template:
            caching: 1
            expires: 3600               # in seconds; use 0 to disable
            cache_stringrefs: 1
            delimiters: [ "{", "}" ]
            prepend: |
                use strict;
                use warnings;
            safe: 1
            safe_opcodes: [ ":default", ":load" ]
            safe_disposable: 0

The following sections explain what these options do.

=head2 Global caching - C<caching>, C<expires>

Contrary to other template engines (like L<Template::Toolkit>), where I<one>
instance may work on I<multiple> templates, I<one> L<Text::Template> instance
is created I<for each> template. Therefore, if:

=for :list
* you don't use a huge amount of different templates;
* you don't use each template just once;

then it may be interesting to B<cache> Text::Template instances for later use.
Since these conditions seem to be common, this engine uses a cache (I<via>
L<CHI>) B<by default>.

If you're OK with caching, you should specify a B<timeout> (C<expires>) after
which cached Text::Template instances are to be refreshed, since you might
have changed your template sources without restarting Dancer2. By default,
this engine uses C<expires: 3600> (one hour). Use C<0> to tell it that
templates never expire.

If you don't want any caching, just set C<caching> to C<0>.

=head2 "String-ref templates" caching - C<cache_stringrefs>

Just like with L<Dancer2::Template::Toolkit>, you can pass templates either as
filenames (for a template file) or string references ("string-refs", which are
dereferenced and used as the template's content). In some cases, you may want
to disable caching for string-refs only: for instance, if you generate a lot
of templates on-the-fly and use them only once, caching them is useless and
fills your cache. You can therefore disable caching I<for string-refs only> by
setting C<cache_stringrefs> to C<0>.

Note that if you set C<caching> to C<0>, you don't have I<any> caching, so
C<cache_stringrefs> is ignored.

=head2 Custom delimiters - C<delimiters>

The C<delimiters> option allows you to specify a custom delimiters pair
(opening and closing) for your templates. See the L<Text::Template>
documentation for more about delimiters, since this module just pass them to
Text::Template. This option defaults to C<{> and C<}>, meaning that in C<< a
{b} c >>, C<b> (and only C<b>) will be interpolated.

=head2 Prepending code - C<prepend>

This option specifies Perl code run by Text::Template I<before> evaluating
each template. For instance, with this option's default value, i.e.:

    use strict;
    use warnings FATAL => 'all';

then evaluating the following template:

    you're the { $a + 1 }th visitor!

is the same as evaluating:

    {
        use strict;
        use warnings FATAL => 'all';
        ""
    }you're the { $a + 1 }th visitor!

and thus you'd get:

    Program fragment delivered error
    ``Use of uninitialized value $a in addition (+) [...]

in your template output if you forgot to pass a value for C<$a>.

If you don't want anything prepended to your templates, simply give a
non-dying, side-effects-free Perl expression to C<prepend>, like C<0> or
C<"">.

=head2 Running in a L<Safe> - C<safe>, C<safe_opcodes>, C<safe_disposable>

This option (enabled by default) makes your templates to be evaluated in a
L<Safe> compartment, i.e. where some potentially dangerous operations (such as
C<system>) are disabled. Note that the same Safe compartment will be used to
evaluate all your templates, unless you explicitly specify C<safe_disposable:
1> (one compartment per template I<evaluation>).

This Safe uses the C<:default> and C<:load> opcode sets (see L<the Opcode
documentation|https://metacpan.org/pod/Opcode#Predefined-Opcode-Tags>), unless
you specify it otherwise with the C<safe_opcodes> option. You can, of course,
mix opcodes and optags, as in:

    safe_opcodes:
        - ":default"
        - "time"

which enables the default opcode set I<and> C<time>. But B<be careful>: with
the previous example for instance, you don't allow C<require>, and thus break
the default value of the C<prepend> option (which contains C<use>)!

=cut

has '+engine' =>
  ( isa => InstanceOf['Dancer2::Template::TextTemplate::FakeEngine'] );

sub _build_engine {
    my $self = shift;
    my $engine = Dancer2::Template::TextTemplate::FakeEngine->new;
    for (qw/ caching expires delimiters cache_stringrefs prepend /) {
        $engine->$_($self->config->{$_}) if $self->config->{$_};
    }
    return $engine;
}

=method render( $template, \%tokens )

Renders the template.

=for :list
* C<$template> is either a (string) filename for the template file or a
  reference to a string that contains the template.
* C<\%tokens> is a hashref for the tokens you wish to pass to
  L<Text::Template> for rendering, as if you were using
  C<Text::Template::fill_in>.

L<Carp|Croak>s if an error occurs.

=cut

sub render {
    my ( $self, $template, $tokens ) = @_;
    $self->engine->process( $template, $tokens )
      or croak $Dancer2::Template::TextTemplate::FakeEngine::ERROR;
}

1;


lib/Dancer2/Template/TextTemplate/FakeEngine.pm
package Dancer2::Template::TextTemplate::FakeEngine;
# ABSTRACT: Fake Template::Toolkit-like, persistent engine around Text::Template.

use strict;
use warnings;

# VERSION

use Moo;
use MooX::Types::MooseLike::Base qw( InstanceOf Bool ArrayRef Int Str );
use Carp 'croak';
use Text::Template 1.46;
use CHI;
use Safe 2.26;
use Scalar::Util 'blessed';
use namespace::clean;

=head1 SYNOPSIS

=head1 DESCRIPTION

With L<Template::Toolkit>-like engines, you instantiate I<one> engine to
process I<multiple> templates.
With L<Text::Template>, you instantiate I<one> engine to process I<one>
template.
This class is a simple wrapper around Text::Template to simplify its use as a
template engine in L<Dancer2>. It basically just manage Text::Template
instances (and their expiration) through L<CHI>.

You can give this engine templates as filenames or string references:

=for :list
* with a filename, the corresponding file will be read by L<Text::Template>;
* a string ref will be dereferenced and its content used as the template
  content itself.

=cut

=for :comment

This is our cache. We should gather CHI-related options from config.yml to
support other caching methods like Memcached. That sounds overkill to me, but
an open interface is always better than a predefined one!

=cut

has _cache => (
    is      => 'rwp',
    isa     => InstanceOf['CHI::Driver'],
    lazy    => 1,
    builder => '_build_cache',
);

sub _build_cache {
    my $self = shift;
    return CHI->new(
        driver             => 'Memory',    # THIS should be generalized
        expires_on_backend => 1,
        global             => 1,           # CHI::Driver::Memory-specific
    );
}

=attr caching

Whether a cache should be used for storing Text::Template instances, or not.

Defaults to C<1> (caching enabled).

=cut

has caching => (
    is      => 'rw',
    isa     => Bool,
    default => 1,
);

=attr expires

How longer (in seconds) the a Text::Template instance will stay in the cache.
After this duration, when the corresponding template is requested again, a new
Text::Template instance will be created (and cached).

Defaults to C<3600> (one hour).
Obviously irrelevant if C<caching> is set to C<0>!

=cut

has expires => (
    is      => 'rw',
    isa     => Int,
    default => 3600,
);

=attr delimiters

An arrayref of (two) delimiters, as defined in L<Text::Template#new>.

Defaults to C<< [ '{', '}' ] >>.

=cut

has delimiters => (
    is      => 'rw',
    isa     => ArrayRef[Str],
    default => sub { [ '{', '}' ] },
);

=attr cache_stringrefs

If this attribute and C<caching> are true (which is the default),
string-ref-templates will always be cached forever (since they cannot become
I<invalid>, contrary to template files that can be changed on disc when
FakeEngine doesn't watch).

However, you may want to disable this behavior for string-ref-templates if you
use a lot of such templates only once (they would fill your cache). By setting
C<cache_stringrefs> to C<0>, you tell FakeEngine not to cache (at all) your
string-ref-templates.

=cut

has cache_stringrefs => (
    is      => 'rw',
    isa     => Bool,
    default => 1,
);

=attr prepend

Contains the string of Perl code added at the top of each evaluated template.
See L<PREPEND in
Text::Template|https://metacpan.org/pod/Text::Template#PREPEND-feature-and-using-strict-in-templates>.

=cut

has prepend => (
    is => 'rw',
    isa => Str,
    default => <<'END',
        use strict;
        use warnings FATAL => 'all';
END
);

# Text::Template's HASH variables (as exclusively used by FakeEngine) are not
# installed in the template evaluation package in a "use strict"-compatible
# way, but we enforce "use strict" in PREPEND, so we need to declare ourselves
# these variables just after the PREPENDed code.
sub _declare_arg_variables {
    my $hash = shift;
    my @decls;
    while ( my ( $name, $value ) = each %$hash ) {
        next unless defined $value;
        push @decls, do {
            if    ( ref $value eq 'ARRAY' ) { '@' }
            elsif ( ref $value eq 'HASH' )  { '%' }
            else                            { '$' }
          }
          . $name;
    }
    return join "\n" => map { "our $_;" } @decls;
}

=attr safe, safe_opcodes, safe_disposable

These attributes are directly linked to the eponymous options in
L<Dancer2::Template::TextTemplate>.

=cut

has safe => (
    is      => 'rw',
    isa     => Bool,
    default => 1,
);

has safe_opcodes => (
    is      => 'rw',
    isa     => ArrayRef[Str],
    default => sub { [qw[ :default require caller dofile ]] },
    trigger => sub {
        my $self = shift;
        $self->_safe->permit_only(@{ $_[0] });
    },
);

has safe_disposable => (
    is      => 'rw',
    isa     => Bool,
    default => 0,
    trigger => sub {
        my $self = shift;
        $self->_rebuild_safe if $_[0];
    },
);

has _safe => (
    is      => 'rw',
    isa     => InstanceOf['Safe'],
    lazy    => 1,
    builder => '_build_safe',
);

sub _build_safe {
    my $self = shift;
    my $safe = Safe->new;
    $safe->permit_only(@{ $self->safe_opcodes });
    return $safe;
}

sub _rebuild_safe {
    my $self = shift;
    $self->_safe($self->_build_safe) if $self->safe_disposable;
    return;
}

=method process( $template, \%tokens )

Computes the C<$template> according to specified C<\%tokens>.

If C<$template> is a string, it is taken as a filename for the template file
(if this file does not exist, the method C<croak>s). If C<$template> is a
scalar reference, it is taken as a reference to a string that contains the
template. In any other case, this method will C<croak> furiously.

Note that, if C<caching> is true, (dereferenced) string references will be
cached too.

This methods simply gets a Text::Template instance (either from cache or by
instantiating it) and calls C<< Text::Template::fill_in( HASH => \%tokens ) >>
on it, returning the result.

If an error occurs in Text::Template, this method returns C<undef>, sets
C<Dancer2::Template::TextTemplate::FakeEngine::ERROR> to the value of
C<Text::Template::ERROR> and does not cache anything.

=cut

sub process {
    my ( $self, $template, $tokens ) = @_;

    defined $template
      && !blessed($template)
      && ( ref $template || -f $template )
      or croak "$template is not a regular file (name) or a string reference";

    my $tt = $self->caching
      ? $self->_cache->get( ref $template ? $$template : $template )
      : undef;
    unless ($tt) {    # either we don't cache or the cached instance expired
        $tt = Text::Template->new(
            TYPE   => ref $template ? 'STRING'   : 'FILE',
            SOURCE => ref $template ? $$template : $template,
            DELIMITERS => $self->delimiters,
        );
    }

    my $computed = $tt->fill_in(
        HASH    => $tokens,
        PREPEND => $self->prepend . _declare_arg_variables($tokens),
        $self->safe ? ( SAFE => $self->_safe ) : (),
    ) or our $ERROR = $Text::Template::ERROR;

    $self->_rebuild_safe if $self->safe && $self->safe_disposable;

    if ( defined $computed && $self->caching ) {
        if ( ref $template && $self->cache_stringrefs ) {
            $self->_cache->set( $$template, $tt, 'never' );
        }
        else {                    # filenames
            $self->_cache->set( $template, $tt,
                $self->expires > 0
                ? { expires_in => $self->expires }
                : 'never' );
        }
    }

    return $computed;
}

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.008_009;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Text::Template engine for Dancer2",
  "AUTHOR" => "Thibaut Le Page <thilp\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Dancer2-Template-TextTemplate",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dancer2::Template::TextTemplate",
  "PREREQ_PM" => {
    "CHI" => 0,
    "Carp" => 0,
    "Dancer2::Core::Role::Template" => 0,
    "Dancer2::Core::Types" => 0,
    "Moo" => 0,
    "MooX::Types::MooseLike::Base" => 0,
    "Safe" => "2.14",
    "Scalar::Util" => 0,
    "Text::Template" => "1.46",
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Temp" => 0,
    "IPC::System::Simple" => 0,
    "Test::API" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "autodie" => 0
  },
  "VERSION" => "1.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CHI" => 0,
  "Carp" => 0,
  "Dancer2::Core::Role::Template" => 0,
  "Dancer2::Core::Types" => 0,
  "File::Temp" => 0,
  "IPC::System::Simple" => 0,
  "Module::Build" => "0.3601",
  "Moo" => 0,
  "MooX::Types::MooseLike::Base" => 0,
  "Safe" => "2.14",
  "Scalar::Util" => 0,
  "Test::API" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.88",
  "Text::Template" => "1.46",
  "autodie" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



